### PR TITLE
Add DPS blob support 

### PR DESCRIPTION
**Pull request by prmathur-microsoft**
## Checklist
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- [ ] This pull-request is submitted against the `master` branch.

### CODE REVIEW

File: provisioning/device/src/DeviceRegistrationResult.cs [hunk 1]
```diff
@@ -43,7 +27,8 @@ public class DeviceRegistrationResult
             DateTime? lastUpdatedDateTimeUtc,
             int errorCode,
             string errorMessage,
-            string etag)
+            string etag,
+            byte[] returnData)
```
Comment (timtay-microsoft) [inline provisioning/device/src/DeviceRegistrationResult.cs]: breaking change. Will need three constructors here, not one

File: provisioning/device/src/Microsoft.Azure.Devices.Provisioning.Client.csproj [hunk 2]
```diff
@@ -17,7 +17,7 @@
   </PropertyGroup>
 
   <PropertyGroup>
-    <Version>1.3.0</Version>
+    <Version>1.2.2</Version>
     <Title>Microsoft Azure IoT Provisioning Device Client SDK</Title>
```
Comment (timtay-microsoft) [inline provisioning/device/src/Microsoft.Azure.Devices.Provisioning.Client.csproj]: Why downgrade?

File: provisioning/device/src/ProvisioningDeviceClient.cs [hunk 3]
```diff
@@ -67,24 +67,29 @@ public class ProvisioningDeviceClient
         /// <summary>
         /// Registers the current device using the Device Provisioning Service and assigns it to an IoT Hub.
         /// </summary>
+        /// <param name="data">The custom content.</param>
         /// <returns>The registration result.</returns>
-        public Task<DeviceRegistrationResult> RegisterAsync()
+        public Task<DeviceRegistrationResult> RegisterAsync(byte[] data)
```
Comment (timtay-microsoft) [inline provisioning/device/src/ProvisioningDeviceClient.cs]: >RegisterAsync [](start = 46, length = 13)

breaking change. Need api for RegisterAsync that takes no parameters

File: provisioning/device/src/ProvisioningErrorDetails.cs [hunk 4]
```diff
@@ -43,14 +42,14 @@ public class ProvisioningErrorDetails
         /// <summary>
         /// Time stamp (in UTC).
         /// </summary>
-        public string TimestampUtc { get; set; }
+        public DateTime TimestampUtc { get; set; }
```
Comment (timtay-microsoft) [inline provisioning/device/src/ProvisioningErrorDetails.cs]: >DateTime [](start = 15, length = 8)

breaking change

File: provisioning/device/src/ProvisioningErrorDetails.cs [hunk 5]
```diff
@@ -43,14 +42,14 @@ public class ProvisioningErrorDetails
...
         /// <summary>
-        /// Create the error message with the saved error code, tracking id and timestamp
+        /// Create details
         /// </summary>
-        /// <param name="message"></param>
-        /// <returns></returns>
-        public string CreateMessage(string message)
+        /// <param name="message">Error message</param>
+        /// <returns>Details</returns>
+        public string CreateDetails(string message)
```
Comment (timtay-microsoft) [inline provisioning/device/src/ProvisioningErrorDetails.cs]: >CreateDetails [](start = 22, length = 13)

breaking change

File: provisioning/device/src/ProvisioningTransportException.cs [hunk 6]
```diff
@@ -88,20 +83,6 @@ public ProvisioningTransportException(string message, Exception innerException,
...
-        /// <param name="isTransient">True if the error is transient.</param>
-        /// <param name="errorDetails">The service error details.</param>
-        /// <param name="innerException">The inner exception.</param>
-        public ProvisioningTransportException(string message, Exception innerException, bool isTransient, ProvisioningErrorDetails errorDetails)
-            : base(message, innerException)
-        {
-            this.IsTransient = isTransient;
-            this.ErrorDetails = errorDetails;
-        }
-
```
Comment (timtay-microsoft) [inline provisioning/device/src/ProvisioningTransportException.cs]: don't remove this

File: provisioning/device/src/ProvisioningTransportException.cs [hunk 7]
```diff
@@ -23,11 +23,6 @@ public class ProvisioningTransportException : Exception
         /// </summary>
         public string TrackingId { get; set; }
 
-        /// <summary>
-        /// Service reported Error details. Use this when reporting a Service issue.
-        /// </summary>
-        public ProvisioningErrorDetails ErrorDetails { get; private set; }
```
Comment (timtay-microsoft) [inline provisioning/device/src/ProvisioningTransportException.cs]: don't remove this

File: provisioning/service/src/Config/AttestationMechanismType.cs [hunk 8]
```diff
@@ -40,12 +40,6 @@ internal enum AttestationMechanismType
         [EnumMember(Value = "x509")]
         X509 = 2,
 
-        /// <summary>
-        /// Symmetric Key attestation mechanism
-        /// </summary>
-        /// <remarks>
-        /// Identify the attestation mechanism as <see cref="SymmetricKeyAttestation"/>
-        /// </remarks>
```
Comment (timtay-microsoft) [inline provisioning/service/src/Config/AttestationMechanismType.cs]: keep this

File: provisioning/service/src/Config/CustomAllocationDefinition.cs [hunk 9]
```diff
@@ -6,22 +6,20 @@
...
     public class CustomAllocationDefinition
     {
         /// <summary>
         /// The webhook URL used for allocation requests.
         /// </summary>
         [JsonProperty(PropertyName = "webhookUrl", Required = Required.Always)]
-#pragma warning disable CA1056 // Uri properties should not be strings
-        public string WebhookUrl { get; set; }
-#pragma warning restore CA1056 // Uri properties should not be strings
+        public string Webhook { get; set; }
```
Comment (timtay-microsoft) [inline provisioning/service/src/Config/CustomAllocationDefinition.cs]: >Webhook [](start = 22, length = 7)

breaking change on property name. Keep as WebhookUrl

File: provisioning/service/src/Config/CustomAllocationDefinition.cs [hunk 10]
```diff
@@ -6,22 +6,20 @@
...
 
         /// <summary>
-        /// The API version of the provisioning service types (such as IndividualEnrollment) sent in the custom allocation request.
+        /// The API version of the provisioning service types (such as IndividualEnrollment) sent in the custom allocation request. Supported versions include: "2018-09-01-preview"
```
Comment (timtay-microsoft) [inline provisioning/service/src/Config/CustomAllocationDefinition.cs]: > Supported versions include: "2018-09-01-preview" [](start = 131, length = 49)

shouldn't recommend preview api versions. 2019-03-31 is the latest

File: provisioning/service/src/Config/EnrollmentGroup.cs [hunk 11]
```diff
@@ -172,14 +172,14 @@ public EnrollmentGroup(string enrollmentGroupId, Attestation attestation)
             AttestationMechanism attestation,
             string iotHubHostName,
             TwinState initialTwinState,
-            ProvisioningStatus? provisioningStatus,
+            ProvisioningStatus provisioningStatus,
```
Comment (timtay-microsoft) [inline provisioning/service/src/Config/EnrollmentGroup.cs]: keep ProvisioningStatus?, not ProvisioningStatus

File: provisioning/service/src/Config/EnrollmentGroup.cs [hunk 12]
```diff
@@ -259,21 +259,14 @@ public Attestation Attestation
             }
             set
```
Comment (timtay-microsoft) [inline provisioning/service/src/Config/EnrollmentGroup.cs]: >set [](start = 12, length = 3)

leave setter alone here. Old code handles exceptions better

File: provisioning/service/src/Config/IndividualEnrollment.cs [hunk 13]
```diff
@@ -22,8 +22,8 @@ namespace Microsoft.Azure.Devices.Provisioning.Service
     /// The minimum information required by the provisioning service is the <code>RegistrationId</code> and the
     /// <code>Attestation</code>.
     ///
-    /// A new device can be provisioned by three attestation mechanisms, Trust Platform Module (see <see cref=
-    /// "TpmAttestation"/>), X509 (see <see cref="X509Attestation"/>) or Symmetric Key (see <see cref="SymmetricKeyAttestation"/>). The definition of each one you 
+    /// A new device can be provisioned by two attestation mechanisms, Trust Platform Module (see <see cref=
+    /// "TpmAttestation"/>) or X509 (see <see cref="X509Attestation"/>). The definition of each one you
```
Comment (timtay-microsoft) [inline provisioning/service/src/Config/IndividualEnrollment.cs]: revert, old documentation includes symmetric key

File: provisioning/service/src/Config/IndividualEnrollment.cs [hunk 14]
```diff
@@ -75,7 +75,7 @@ public class IndividualEnrollment : IETagHolder
         /// This constructor creates an instance of the IndividualEnrollment object with the minimum set of 
         /// information required by the provisioning service. A valid individualEnrollment must contain the 
         /// registrationId, which uniquely identify this enrollment, and the attestation mechanism, which can 
-        /// be TPM, X509, or Symmetric key.
+        /// be TPM or X509.
```
Comment (timtay-microsoft) [inline provisioning/service/src/Config/IndividualEnrollment.cs]: revert

File: provisioning/service/src/Config/SymmetricKeyAttestation.cs [hunk 15]
```diff
@@ -7,20 +7,27 @@
 namespace Microsoft.Azure.Devices.Provisioning.Service
 {
     /// <summary>
-    /// Attestation using a symmetric key
+    /// Symmetric key
```
Comment (timtay-microsoft) [inline provisioning/service/src/Config/SymmetricKeyAttestation.cs]: revert

File: provisioning/service/src/Config/SymmetricKeyAttestation.cs [hunk 16]
```diff
@@ -35,6 +42,10 @@ public string PrimaryKey
             }
             private set
             {
+                if (!string.IsNullOrEmpty(value))
+                {
+                    ParserUtils.EnsureBase64String(value);
+                }
```
Comment (timtay-microsoft) [inline provisioning/service/src/Config/SymmetricKeyAttestation.cs]: don't check this, let the server check it instead

File: provisioning/service/src/Config/SymmetricKeyAttestation.cs [hunk 17]
```diff
@@ -52,9 +63,13 @@ public string SecondaryKey
...
```
Comment (timtay-microsoft) [inline provisioning/service/src/Config/SymmetricKeyAttestation.cs]: don't check this, let the server check it instead

File: provisioning/service/src/Config/SymmetricKeyAttestation.cs [hunk 18]
```diff
@@ -7,20 +7,27 @@
...
-            SecondaryKey = secondaryKey;
+            try
+            {
+                PrimaryKey = primaryKey;
+                SecondaryKey = secondaryKey;
+            }
+            catch (ArgumentException e)
+            {
+                throw new ProvisioningServiceClientException(e);
+            }
```
Comment (timtay-microsoft) [inline provisioning/service/src/Config/SymmetricKeyAttestation.cs]: try catch not needed once you stop verifying base64 string

File: provisioning/service/src/Config/TpmAttestation.cs [hunk 19]
```diff
@@ -56,12 +56,9 @@ public string EndorsementKey
...
             {
-                if (string.IsNullOrWhiteSpace(value))
-                {
-                    /* SRS_TPM_ATTESTATION_21_001: [The EndorsementKey setter shall throws ArgumentNullException if the provided 
-                                                endorsementKey is null or white space.] */
-                    throw new ArgumentNullException(nameof(value));
-                }
+                /* SRS_TPM_ATTESTATION_21_001: [The EndorsementKey setter shall throws ArgumentException if the provided 
+                                                endorsementKey is null or invalid.] */
+                ParserUtils.EnsureBase64String(value);
```
Comment (timtay-microsoft) [inline provisioning/service/src/Config/TpmAttestation.cs]: >ParserUtils.EnsureBase64String(value); [](start = 16, length = 38)

don't check this, let the server check it instead

File: provisioning/service/src/Config/TpmAttestation.cs [hunk 20]
```diff
@@ -79,7 +76,12 @@ public string StorageRootKey
             }
             private set
             {
-                /* SRS_TPM_ATTESTATION_21_002: [The StorageRootKey setter shall store the storageRootKey passed.] */
+                /* SRS_TPM_ATTESTATION_21_002: [The StorageRootKey setter shall throws ArgumentException if the provided 
+                                                storageRootKey is not null but invalid.] */
+                if (value != null)
+                {
+                    ParserUtils.EnsureBase64String(value);
+                }
```
Comment (timtay-microsoft) [inline provisioning/service/src/Config/TpmAttestation.cs]: don't check this, let the server check it instead

File: provisioning/service/src/Config/ParserUtility.cs [hunk 21]
```diff
@@ -44,6 +44,34 @@ public static void EnsureUTF8String(string input)
...
+            {
+                Convert.FromBase64String(input);
+            }
+            catch (FormatException e)
+            {
+                throw new ArgumentException("parameter is not a valid Base64 string", e);
+            }
+
+            /* Codes_SRS_PARSER_UTILITY_21_004: [The IsValidBase64 shall do nothing if the string is valid.] */
+        }
```
Comment (timtay-microsoft) [inline provisioning/service/src/Config/ParserUtility.cs]: Not needed, only server should be checking this

File: provisioning/service/src/Contract/SDKUtils.cs [hunk 22]
```diff
@@ -5,7 +5,8 @@ namespace Microsoft.Azure.Devices.Provisioning.Service
 {
     internal class SDKUtils
     {
-        private const string ApiVersionProvisioning = "2018-11-01";
-        public const string ApiVersionQueryString = CustomHeaderConstants.ApiVersion + "=" + ApiVersionProvisioning;
+        // TODO - Since there are breaking changes on Attestation and other data contracts in 2019-01-15 version, can't update to the latest api version until all these changes applied on.
+        private const string ApiVersionProvisioningPreview = "2019-02-15-preview";
+        public const string ApiVersionQueryString = CustomHeaderConstants.ApiVersion + "=" + ApiVersionProvisioningPreview;
```
Comment (timtay-microsoft) [inline provisioning/service/src/Contract/SDKUtils.cs]: leave as 2018-11-01

Comment (timtay-microsoft) [reply provisioning/service/src/Contract/SDKUtils.cs]: My DeviceCapabilities for enrollmentGroup feature will bump this up to 2019-03-31

---
In reply to: [274179262](https://github.com/Azure/azure-iot-sdk-csharp/pull/905#discussion_r274179262) [](ancestors = 274179262)

File: provisioning/service/tests/Config/AttestationMechanismTests.cs [hunk 23]
```diff
@@ -212,36 +213,9 @@ public void AttestationMechanismConstructorJSONSucceedOnNoneType()
...
-            
-            //assert
-            Assert.IsNotNull(attestationMechanism);
-            Assert.IsTrue(attestationMechanism.Type == AttestationMechanismType.SymmetricKey);
-            Assert.IsTrue(attestationMechanism.GetAttestation() is SymmetricKeyAttestation);
-            SymmetricKeyAttestation symmetricKeyAttestation = (SymmetricKeyAttestation) attestationMechanism.GetAttestation();
-
-            Assert.AreEqual(samplePrimaryKey, symmetricKeyAttestation.PrimaryKey);
-            Assert.AreEqual(sampleSecondaryKey, symmetricKeyAttestation.SecondaryKey);
-        }
```
Comment (timtay-microsoft) [inline provisioning/service/tests/Config/AttestationMechanismTests.cs]: still a valid test, don't delete

File: provisioning/service/tests/Config/ParserUtilsTests.cs [hunk 24]
```diff
@@ -30,6 +30,28 @@ public void ParserUtilsIsValidUTF8ThrowsOnInvalidUTF8()
...
+        [TestMethod]
+        public void ParserUtils_IsValidBase64_ThrowsOnInvalidBase64()
+        {
+            // arrange - act - assert
+            TestAssert.Throws<ArgumentException>(() => ParserUtils.EnsureBase64String(null));
+            TestAssert.Throws<ArgumentException>(() => ParserUtils.EnsureBase64String(""));
+            TestAssert.Throws<ArgumentException>(() => ParserUtils.EnsureBase64String("  "));
+            TestAssert.Throws<ArgumentException>(() => ParserUtils.EnsureBase64String("thisisnotavalidbase64="));
+            TestAssert.Throws<ArgumentException>(() => ParserUtils.EnsureBase64String("this is not a valid UTF8 \u1234"));
+        }
```
Comment (timtay-microsoft) [inline provisioning/service/tests/Config/ParserUtilsTests.cs]: won't be necessary when you stop testing for base64 strings

File: provisioning/shared/src/device/provisioning/transport/ClientApiVersionHelper.cs [hunk 25]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.Devices.Provisioning.Client
+{
+    internal class ClientApiVersionHelper
+    {
+        // TODO: Split ApiVersionName as it is only used by HTTP and AMQP.
+        public const string ApiVersionName = "api-version";
+        public const string January2019ApiVersion = "2019-01-15";
```
Comment (timtay-microsoft) [inline provisioning/shared/src/device/provisioning/transport/ClientApiVersionHelper.cs]: >January2019ApiVersion [](start = 28, length = 21)

We don't want to be using this anywhere, do we?

File: provisioning/shared/src/device/provisioning/transport/ProvisioingSasBuilder.cs [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+using System.Collections.Generic;
+using System.Globalization;
+using System.Net;
+using System.Text;
+using Microsoft.Azure.Devices.Shared;
+
+namespace Microsoft.Azure.Devices.Provisioning.Client.Transport
+{
+    //TODO: merge with other Sas Token Builder in IoTHubClient
+    internal static class ProvisioningSasBuilder
```
Comment (timtay-microsoft) [inline provisioning/shared/src/device/provisioning/transport/ProvisioingSasBuilder.cs]: >ProvisioningSasBuilder [](start = 26, length = 22)

should not be needed. This class already exists elsewhere

File: provisioning/shared/src/DefaultWebProxySettings.cs [hunk 27]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (davilu) [inline provisioning/shared/src/DefaultWebProxySettings.cs]: Are those new files duplicated from shared?

File: provisioning/device/src/ProductInfo.cs [hunk 28]
```diff
@@ -16,12 +15,15 @@ public override string ToString()
         {
             const string Name = "Microsoft.Azure.Devices.Provisioning.Client";
 
-            string version = typeof(ProvisioningDeviceClient).GetTypeInfo().Assembly.GetName().Version.ToString(3);
+            // TODO: Replace with Assembly information.
+            // DO NOT EDIT the following line; it is updated by the bump_version script (https://github.com/Azure/iot-sdks-internals/blob/master/release/csharp/inputs.js)
+            const string Version = "1.0.2"; // CommonAssemblyVersion
+
             string runtime = RuntimeInformation.FrameworkDescription.Trim();
```
Comment (prmathur-microsoft) [inline provisioning/device/src/ProductInfo.cs]: @davilu  - Why is this editing needed?

File: provisioning/device/src/ProductInfo.cs [hunk 29]
```diff
@@ -16,12 +15,15 @@ public override string ToString()
...
             string operatingSystem = RuntimeInformation.OSDescription.Trim();
             string processorArchitecture = RuntimeInformation.ProcessArchitecture.ToString().Trim();
 
-            string userAgent = $"{Name}/{version} ({runtime}; {operatingSystem}; {processorArchitecture})";
+            string userAgent = $"{Name}/{Version} ({runtime}; {operatingSystem}; {processorArchitecture})";
```
Comment (prmathur-microsoft) [inline provisioning/device/src/ProductInfo.cs]: Why do we need to change this? We should not touch useragent

File: provisioning/device/src/ProvisioningDeviceClient.cs [hunk 30]
```diff
@@ -67,24 +67,29 @@ public class ProvisioningDeviceClient
...
 
         /// <summary>
         /// Registers the current device using the Device Provisioning Service and assigns it to an IoT Hub.
         /// </summary>
+        /// <param name="data">The custom content.</param>
         /// <param name="cancellationToken">The cancellation token.</param>
         /// <returns>The registration result.</returns>
-        public Task<DeviceRegistrationResult> RegisterAsync(CancellationToken cancellationToken)
+        public Task<DeviceRegistrationResult> RegisterAsync(byte[] data, CancellationToken cancellationToken)
         {
```
Comment (prmathur-microsoft) [inline provisioning/device/src/ProvisioningDeviceClient.cs]: Breaking change. Modifying a public API

File: provisioning/device/src/ProvisioningDeviceClient.cs [hunk 31]
```diff
@@ -67,24 +67,29 @@ public class ProvisioningDeviceClient
...
             if (Logging.IsEnabled) Logging.RegisterAsync(this, _globalDeviceEndpoint, _idScope, _transport, _security);
 
-            var request = new ProvisioningTransportRegisterMessage(_globalDeviceEndpoint, _idScope, _security);
-            request.ProductInfo = ProductInfo;
+            var request = new ProvisioningTransportRegisterMessage(_globalDeviceEndpoint, _idScope, _security, data)
+            {
+                ProductInfo = ProductInfo
+            };
             return _transport.RegisterAsync(request, cancellationToken);
```
Comment (prmathur-microsoft) [inline provisioning/device/src/ProvisioningDeviceClient.cs]: Why is this needed?

File: provisioning/device/src/ProvisioningErrorDetails.cs [hunk 32]
```diff
@@ -1,17 +1,16 @@
...
 using System.Text;
+using Newtonsoft.Json;
 
 namespace Microsoft.Azure.Devices.Provisioning.Client
 {
     /// <summary>
     /// Provisioning error details.
     /// </summary>
-    [SuppressMessage("Microsoft.Performance", "CA1812", Justification = "Used by the JSon parser.")]
     public class ProvisioningErrorDetails
```
Comment (prmathur-microsoft) [inline provisioning/device/src/ProvisioningErrorDetails.cs]: Why remove this?

File: provisioning/device/src/ProvisioningTransportRegisterMessage.cs [hunk 33]
```diff
@@ -0,0 +1,74 @@
...
+using Newtonsoft.Json.Linq;
+using System.Text;
+
+namespace Microsoft.Azure.Devices.Provisioning.Client.Transport
+{
+    /// <summary>
+    /// Represents a Provisioning registration message.
+    /// </summary>
+    public class ProvisioningTransportRegisterMessage
+    {
```
Comment (prmathur-microsoft) [inline provisioning/device/src/ProvisioningTransportRegisterMessage.cs]: Why is this class needed?

Comment (prmathur-microsoft) [reply provisioning/device/src/ProvisioningTransportRegisterMessage.cs]: Can't Data be added to existing?

File: provisioning/service/src/Auth/ServiceConnectionString.cs [hunk 34]
```diff
@@ -23,6 +23,7 @@ internal sealed class ServiceConnectionString : IAuthorizationHeaderProvider
     {
         private static readonly TimeSpan DefaultTokenTimeToLive = TimeSpan.FromHours(1);
         private const string UserSeparator = "@";
+        private const int DefaultHttpsPort = 443;
```
Comment (prmathur-microsoft) [inline provisioning/service/src/Auth/ServiceConnectionString.cs]: Required to default to WS?

File: provisioning/service/src/Auth/ServiceConnectionStringBuilder.cs [hunk 35]
```diff
@@ -21,7 +21,7 @@ internal class ServiceConnectionStringBuilder
         private static readonly string SharedAccessKeyNamePropertyName = ((MemberExpression)((Expression<Func<ServiceConnectionStringBuilder, string>>)(_ => _.SharedAccessKeyName)).Body).Member.Name; // todo: replace with nameof()
         private static readonly string SharedAccessKeyPropertyName = ((MemberExpression)((Expression<Func<ServiceConnectionStringBuilder, string>>)(_ => _.SharedAccessKey)).Body).Member.Name; // todo: replace with nameof()
         private static readonly string SharedAccessSignaturePropertyName = ((MemberExpression)((Expression<Func<ServiceConnectionStringBuilder, string>>)(_ => _.SharedAccessSignature)).Body).Member.Name; // todo: replace with nameof();
-
+        static readonly string PortPropertyName = ((MemberExpression)((Expression<Func<ServiceConnectionStringBuilder, int?>>)(_ => _.Port)).Body).Member.Name; // todo: replace with nameof();
         private static readonly TimeSpan regexTimeoutMilliseconds = TimeSpan.FromMilliseconds(500);
```
Comment (prmathur-microsoft) [inline provisioning/service/src/Auth/ServiceConnectionStringBuilder.cs]: Handle TODO

File: provisioning/device/src/ProductInfo.cs (hunk 29 above)
Comment (davilu) [reply provisioning/device/src/ProductInfo.cs]: Please see the comments of Version on right above. It's from Sam, not sure what it is doing.

File: provisioning/service/src/Config/AllocationPolicy.cs [hunk 36]
```diff
@@ -12,25 +14,25 @@ namespace Microsoft.Azure.Devices.Provisioning.Service
...
-        /// Use a static allocation
+        /// Shard by static setting
         /// </summary>
         [EnumMember(Value = "static")]
         Static = 2,
 
         /// <summary>
-        /// Use a custom allocation
+        /// Shard by static customized setting
         /// </summary>
```
Comment (prmathur-microsoft) [inline provisioning/service/src/Config/AllocationPolicy.cs]: Revert these comments

File: provisioning/service/src/Config/AttestationMechanism.cs [hunk 37]
```diff
@@ -24,7 +24,7 @@ internal sealed class AttestationMechanism
         /// <remarks>
         /// It will create a new instance of the AttestationMechanism for the provided attestation type.
         /// </remarks>
-        /// <param name="attestation">the <see cref="Attestation"/> with the TPM keys, X509 certificates, or Symmetric keys. It cannot 
+        /// <param name="attestation">the <see cref="Attestation"/> with the TPM keys or X509 certificates. It cannot
```
Comment (prmathur-microsoft) [inline provisioning/service/src/Config/AttestationMechanism.cs]: Revert the comment

File: provisioning/service/src/Config/CustomAllocationDefinition.cs [hunk 38]
```diff
@@ -6,22 +6,20 @@
 namespace Microsoft.Azure.Devices.Provisioning.Service
 {
     /// <summary>
-    /// Defines a custom allocation
+    /// Function link details 
     /// </summary>
```
Comment (prmathur-microsoft) [inline provisioning/service/src/Config/CustomAllocationDefinition.cs]: Revert

File: provisioning/service/src/Config/DeviceRegistrationState.cs [hunk 39]
```diff
@@ -74,6 +74,12 @@ public DeviceRegistrationState(string registrationId)
...
         public string ETag { get; internal set; }
-    }
 
+
+        /// <summary>
+        /// Custom data returned from the webhook to the device..
+        /// </summary>
+        [JsonProperty(PropertyName = "returnData", DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public TwinCollection ReturnData { get; internal set; }
+    }
```
Comment (prmathur-microsoft) [inline provisioning/service/src/Config/DeviceRegistrationState.cs]: Is this twin?

File: provisioning/service/src/Microsoft.Azure.Devices.Provisioning.Service.csproj [hunk 40]
```diff
@@ -18,7 +18,7 @@
   </PropertyGroup>
 
   <PropertyGroup>
-    <Version>1.4.0</Version>
+    <Version>1.3.1</Version>
     <Title>Microsoft Azure IoT Provisioning Service Client SDK</Title>
```
Comment (prmathur-microsoft) [inline provisioning/service/src/Microsoft.Azure.Devices.Provisioning.Service.csproj]: Revert to new version

File: provisioning/service/src/Microsoft.Azure.Devices.Provisioning.Service.csproj [hunk 41]
```diff
@@ -31,27 +31,24 @@
...
     <PackageReference Condition=" '$(Configuration)' == 'Debug' " Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.6.1" />
   </ItemGroup>
 
-  <ItemGroup>
-    <ProjectReference Include="$(RootDir)\shared\src\Microsoft.Azure.Devices.Shared.csproj" />
-  </ItemGroup>
-
   <ItemGroup>
     <Compile Update="Properties\ApiResources.Designer.cs">
       <DesignTime>True</DesignTime>
```
Comment (prmathur-microsoft) [inline provisioning/service/src/Microsoft.Azure.Devices.Provisioning.Service.csproj]: Which is the right one to pick up here?

File: provisioning/shared/src/device/provisioning/transport/DeviceRegistrationResultConvertor.cs [hunk 42]
```diff
@@ -0,0 +1,37 @@
...
+using Newtonsoft.Json;
+using System;
+
+namespace Microsoft.Azure.Devices.Provisioning.Client.Transport
+{
+    internal class DeviceRegistrationResultConvertor
+    {
+        internal static DeviceRegistrationResult ConvertToProvisioningRegistrationResult(RegistrationResult result)
+        {
+            var status = ProvisioningRegistrationStatusType.Failed;
```
Comment (prmathur-microsoft) [inline provisioning/shared/src/device/provisioning/transport/DeviceRegistrationResultConvertor.cs]: Why si this conversion needed? Can't it sit in same class?

File: provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs [hunk 43]
```diff
@@ -162,7 +164,7 @@ private Task ConnectAsync(IChannelHandlerContext context)
                     UsernameFormat,
                     _message.IdScope,
                     registrationId,
-                    ClientApiVersionHelper.ApiVersion,
+                    ClientApiVersionHelper.January2019ApiVersion,
                     Uri.EscapeDataString(userAgent)),
```
Comment (timtay-microsoft) [inline provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs]: Don't use "January2019ApiVersion", just use latest version

File: provisioning/transport/http/src/SymmetricKeyCredentials.cs [hunk 44]
```diff
@@ -32,15 +32,59 @@ public override Task ProcessHttpRequestAsync(HttpRequestMessage request, Cancell
...
+            return buffer.ToString();
+        }
+
+        private static string Sign(string requestString, string key)
+        {
+            using (var hmac = new HMACSHA256(Convert.FromBase64String(key)))
+            {
+                return Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(requestString)));
+            }
         }
```
Comment (timtay-microsoft) [inline provisioning/transport/http/src/SymmetricKeyCredentials.cs]: Duplicate code from elsewhere, combine in ProvisioningSasBuilder

File: provisioning/transport/http/src/ApiVersionDelegatingHandler.cs [hunk 45]
```diff
@@ -12,12 +12,19 @@ namespace Microsoft.Azure.Devices.Provisioning.Client.Transport
 {
     internal class ApiVersionDelegatingHandler : DelegatingHandler
     {
+        string _apiVersion;
+
+        public ApiVersionDelegatingHandler(string apiVersion = ClientApiVersionHelper.January2019ApiVersion)
+        {
+            _apiVersion = apiVersion;
+        }
+
```
Comment (timtay-microsoft) [inline provisioning/transport/http/src/ApiVersionDelegatingHandler.cs]: Don't need this. We should always use the latest api version

File: provisioning/transport/amqp/src/ProvisioningTransportHandlerAmqp.cs [hunk 46]
```diff
@@ -155,31 +153,41 @@ public class ProvisioningTransportHandlerAmqp : ProvisioningTransportHandler
...
-            amqpReceivingLink.AddApiVersion(ClientApiVersionHelper.ApiVersion);
+            amqpReceivingLink.AddApiVersion(apiVersion);
 
             await amqpReceivingLink.OpenAsync(TimeoutConstant).ConfigureAwait(false);
 
             var amqpSendingLink = amqpDeviceSession.CreateSendingLink(linkEndpoint);
 
             amqpSendingLink.AddClientVersion(productInfo);
-            amqpSendingLink.AddApiVersion(ClientApiVersionHelper.ApiVersion);
+            amqpSendingLink.AddApiVersion(apiVersion);
```
Comment (timtay-microsoft) [inline provisioning/transport/amqp/src/ProvisioningTransportHandlerAmqp.cs]: >apiVersion [](start = 42, length = 10)

Just use the latest api version, here and above

File: provisioning/transport/amqp/src/AmqpAuthStrategySymmetricKey.cs [hunk 47]
```diff
@@ -35,7 +35,7 @@ public override AmqpSettings CreateAmqpSettings(string idScope)
             SaslPlainHandler saslHandler = new SaslPlainHandler();
             saslHandler.AuthenticationIdentity = $"{idScope}/registrations/{_security.GetRegistrationID()}";
             string key = _security.GetPrimaryKey();
-            saslHandler.Password = ProvisioningSasBuilder.BuildSasSignature("registration", key, saslHandler.AuthenticationIdentity, TimeSpan.FromDays(1));
+            saslHandler.Password = BuildSasSignature("registration", key, saslHandler.AuthenticationIdentity, TimeSpan.FromDays(1));
```
Comment (timtay-microsoft) [inline provisioning/transport/amqp/src/AmqpAuthStrategySymmetricKey.cs]: Use the ProvisioningSasBuilder here, don't duplicate its code in this class

**Comment by CIPop** [comment]
Closed per offline discussion with @prmathur-microsoft .
